### DISCUSSION TITLE
Power On Event?

### DISCUSSION THREAD

**Discussion by lcnittl** [PiSupply/PiJuice#701] [Q&A] [answered]
I was wondering if there is an ~`power on`~ `power back` event that works like ~`power off`~ `no power` but is triggered when external energy supply is available again?

**Comment by tvoverbeek** [comment]
Not directly.
There is 'Wakeup On Charge' which you could use with the level set at 0 or 1%

**Reply by lcnittl** [reply]
Thanks for your reply!

Hmm, I see. So this would allow to wake the pi up if it went to sleep, right? But is it also possible to link it with an event/user script? Would I be able to execute a script when the pi is not yet powered down?

My use scenario for the `power on` event would be as an example: `power off` event is triggered as power supply stopped - pi informs me via email or similar. Knowing that the battery will last for some time (the sent email contains , let's say, charging info), I now plan at what time I can be at the pi to fix power problems. Now, if power comes back in the meantime (the pi has not been powered down so far), it would be nice if it could send me another message that is triggered by a `power on` type event.

Would be cool if something like this could be implemented in the PiJuice Server.

Worst case (if not gets implemented, and also for now), I can have the script that is executed on `power off` check periodically if power is back via PiJuice  API, and then report. Yet, a built in event feels much nicer somehow.

**Comment by tvoverbeek** [comment]
Just execute a job at boot (e.g. using a cron @reboot job) which checks if external power exists and if so sends you an email
(after the network is up).
To avoid emails at 'normal' boots let your no power function store an empty file somewhere as a signal and the reboot job check for this file, then deletes it and only then send you an email.

**Reply by lcnittl** [reply]
Totally makes sense if the pi was shut down already - in that case also an event would not make sense because no one would be there to listen to. I would require to detect `power on`, however, in a state when the pi is still up and running.

As I mentioned above, I could implement repeated checks in the script executed by the `power off` event, however, regarding energy efficiency it might be better to have PiJuice server detect this and execute a new script (as PiJuice server is anyways running in the background).

**Comment by tvoverbeek** [comment]
I do not see the problem. If the 'no power' event is not triggered (and the Pi is running) there must be external power

**Reply by lcnittl** [reply]
I have the feeling we are talking past each other, sorry for that. I will try to reiterate my scenario:

1. Pi with PiJuice is running on external power, battery is fully charged.
2. A power interruption triggers a `no power` event, and PiJuice executes a user script (which informs me of the power shortage via e.g. email). This script does not shut down the Pi. Shutting down is left to `low battery voltage` and/or `low charge` events in the case that external power supply is not back before the *juice* runs out.
3. The power interruption is over, the Pi is receiving external power again (it has been running on battery in the meantime). Triggering of a (not (yet?) implemented) `power on` event executes another user script (which informs me that power supply is back so that I do not have worry anymore).

I hope this helped to clarify a bit. Thanks a lot

**Comment by tvoverbeek** [comment]
Yes, this clarified it. I missed the 'not shutting down' when the 'no power' event occurs.
I added a 'power restored' event to my own setup and it seems to work as you want.
Will do some more testing and then add the modified pijuice_sys.py (the service) to this issue for testing (hopefully tomorrow).

**Reply by lcnittl** [reply]
Excellent – Thanks a lot in advance! Looking forward to testing it.

PS: Seems like I was clever enough to mistype `no power` as `power off`, which certainly did not help in conveying my matter :-).

**Comment by tvoverbeek** [comment, marked as answer]
@lcnittl Attached the modified pijuice_sys.py and a modified pijuice_cli.py (to manipulate the added 'power back' setting)
[pijuice_cli.py.txt](https://github.com/PiSupply/PiJuice/files/6303716/pijuice_cli.py.txt)
[pijuice_sys.py.txt](https://github.com/PiSupply/PiJuice/files/6303717/pijuice_sys.py.txt)
Remove the .txt extension (was needed because github does not allow uploading .py files)

To test:
- Back up the existing /usr/bin/pijuice_sys.py
- Stop the pijuice service `sudo systemctl stop pijuice`
- Copy the modified pijuice_sys.py to /usr/bin (needs sudo)
- Set correct owner `sudo chown pijuice:pijuice /usr/bin/pijuice_sys.py`
- Start the pijuice service again `sudo systemctl start pijuice` and check with `systemctl status pijuice`
- You can run the changed pijuice_cli.py with `sudo -u pijuice python3 <path to modified>pijuice_cli.py` to
  enable the new 'power back' option

Notes: the 'power back' option only works when 'no power' is enabled.
To test I linked the 'no power' and 'power back' events to USER_FUNC1.
For USER_FUNC1 i used the user_func1.py from https://github.com/PiSupply/PiJuice/blob/master/Software/Test/user_func1.py which logs events to a file.

Pleae test and let me go how it goes. If no major problems i will commit it to the main branch so it will appear in the next version.

**Reply by lcnittl** [reply]
Yes, that works as expected – Thanks for the fast implementation!

May I ask why you link triggering of `power back` event to the enabled flag of `no power`? I believe that this is a not necessary limitation and might also make the configuration a bit in-transparent (unless `power back` option is "grayed out" whenever `no power` is not checked). Plus, why not allow the user to only get informed when power is back without forcing to enable the `no power` event and link it to `NO_FUNC`. You could circumvent this by changing the following in `pijuice_sys.py`, if I am not mistaken, this should not have unwanted side effects:

```py
# L40
PowCnt = 0

# L47 - Inserted line
NO_POWER_STATUSES = ['NOT_PRESENT', 'BAD']

# L189
def _EvalPowerInputs(status):
    if (status['battery'] == 'NOT_PRESENT'): return
    global noPowCnt, PowCnt
    print(f"noPowCnt = {noPowCnt}, PowCnt = {PowCnt}")  # [lcnittl] Please remove me :D
    if status['powerInput'] in NO_POWER_STATUSES and status['powerInput5vIo'] in NO_POWER_STATUSES:
        # power is absent
        noPowCnt = min(noPowCnt + 1, 3)
        if noPowCnt == 2:
            if noPowEn:
                print("Triggered noPowEn")  # [lcnittl] Please remove me :D
                # unplugged
                ExecuteFunc(configData['system_events']['no_power']['function'],
                                       'no_power', '')
            PowCnt = 0 # enable checking for return of power
    else:
        # power is present
        PowCnt = min(PowCnt + 1, 3)
        if PowCnt == 2:
            if PowEn and noPowCnt >= 2:
                print("Triggered PowEn")  # [lcnittl] Please remove me :D
                ExecuteFunc(configData['system_events']['power']['function'],
                                       'power', '')
            noPowCnt = 0

# L418-419
                    if noPowEn or PowEn:
                        _EvalPowerInputs(status)
```

For simplicity reasons I, please find the full `pijuice_sys.py` file here: [pijuice_sys.py.txt](https://github.com/PiSupply/PiJuice/files/6306212/pijuice_sys.py.txt) (Same file-extension hack + Please remove the `# [lcnittl] comments`.

Which gives the following output:

```py
RTC os-support OK
noPowCnt = 0, PowCnt = 0
noPowCnt = 0, PowCnt = 1
noPowCnt = 0, PowCnt = 2
noPowCnt = 0, PowCnt = 3
noPowCnt = 0, PowCnt = 3
# [...]
noPowCnt = 0, PowCnt = 3
noPowCnt = 0, PowCnt = 3
# Turned off external power
noPowCnt = 1, PowCnt = 3
Triggered noPowEn
noPowCnt = 2, PowCnt = 0
noPowCnt = 3, PowCnt = 0
noPowCnt = 3, PowCnt = 0
# [...]
noPowCnt = 3, PowCnt = 0
noPowCnt = 3, PowCnt = 0
# Turned on external power
noPowCnt = 3, PowCnt = 1
Triggered PowEn
noPowCnt = 0, PowCnt = 2
noPowCnt = 0, PowCnt = 3
noPowCnt = 0, PowCnt = 3
# [...]
noPowCnt = 0, PowCnt = 3
noPowCnt = 0, PowCnt = 3
# Turned off external power
noPowCnt = 1, PowCnt = 3
Triggered noPowEn
noPowCnt = 2, PowCnt = 0
# Turned on external power
noPowCnt = 2, PowCnt = 1
Triggered PowEn
noPowCnt = 0, PowCnt = 2
# Turned off external power
noPowCnt = 1, PowCnt = 2
Triggered noPowEn
noPowCnt = 2, PowCnt = 0
# Turned on external power
noPowCnt = 2, PowCnt = 1
Triggered PowEn
noPowCnt = 0, PowCnt = 2
noPowCnt = 0, PowCnt = 3
noPowCnt = 0, PowCnt = 3
# [...]
```

If you want the `no power` event not to be triggered on service start with missing external power, you could add the same logic from L209 (my file) that prevents firing of the event at service start with present external power to L199 (my file), thus being

```py
if noPowEn and PowCnt >= 2:
```

Although I am quite sure you want to trigger `no power` in any case, no only if it occurs while service is already running.

Sorry for this rather unconventional "pull request" and for messing with your code :-D

LG nach Tirol!

**Reply by lcnittl** [reply]
> You could circumvent this by changing the following in pijuice_sys.py, if I am not mistaken, this should not have unwanted side effects:

Well, turns out I was mistaken...

The problem was, that both `PowCnt` and `noPowCnt` accumulated if switched at state `1`. Thus, 1 short power break did not trigger the event, but `noPowCnt` was set to `1` until eventually another short cut happened setting it to `2` and triggering the event.

I have now fixed this, essentially by setting `PwrCnt` initially to `3` (set it to `0` if you want to trigger on service start with power supply on - set both to `3` to deactivate both event triggers at service start) and resetting the respectively other counter as soon as the first one was increased in the previous loop (I hope that is somewhat clear :) ).

```py
# L40
PowCnt = 3

# L191
def _EvalPowerInputs(status):
    if (status['battery'] == 'NOT_PRESENT'): return
    global noPowCnt, PowCnt
    print(f"noPowCnt = {noPowCnt}, PowCnt = {PowCnt}")  # [lcnittl] Please remove me :D
    if status['powerInput'] in NO_POWER_STATUSES and status['powerInput5vIo'] in NO_POWER_STATUSES:
        # power is absent
        if noPowCnt:
            PowCnt = 0 # enable checking for return of power
        noPowCnt = min(noPowCnt + 1, 3)
        if noPowEn and noPowCnt == 2:
            print("Triggered noPowEn")  # [lcnittl] Please remove me :D
            # unplugged
            ExecuteFunc(configData['system_events']['no_power']['function'],
                                   'no_power', '')
    else:
        # power is present
        if PowCnt:
            noPowCnt = 0
        PowCnt = min(PowCnt + 1, 3)
        if PowEn and PowCnt == 2:
            print("Triggered PowEn")  # [lcnittl] Please remove me :D
            ExecuteFunc(configData['system_events']['power']['function'],
                                   'power', '')
```

This now successfully resets the counters from `1` to `0`:

```py
# Turned off external power
RTC os-support OK
noPowCnt = 0, PowCnt = 3
noPowCnt = 1, PowCnt = 3
Triggered noPowEn
noPowCnt = 2, PowCnt = 0
noPowCnt = 3, PowCnt = 0
noPowCnt = 3, PowCnt = 0
noPowCnt = 3, PowCnt = 0
# Turned on external power
noPowCnt = 3, PowCnt = 1
# Turned off external power
noPowCnt = 3, PowCnt = 0

```

Here is the complete file: [pijuice_sys.py.txt](https://github.com/PiSupply/PiJuice/files/6306557/pijuice_sys.py.txt)

**Reply by lcnittl** [reply]
If you prefer, I can also make a real PR for that.

**Reply by lcnittl** [reply]
One last thing I would propose to change in my script in order to slightly cut calculation amount is to change
```py
noPowCnt = min(noPowCnt + 1, 3)
```
to either
```py
noPowCnt += 1 if noPowCnt < 3 else 0
```
or
```py
if noPowCnt < 3:
    noPowCnt += 1
```
for both `noPowCnt ` and `PowCnt `, respectively. This removes the calculation overhead of always summing numbers before comparing.

The last post - I promise 😁

**Reply by tvoverbeek** [reply]
I tested your last version and it works like it should. 'power' and 'no power' events are now independent.
'no power' is triggered at boot while 'power' is not.
Thanks/Vielen Dank for your changes.
I'll still have to add the 'power' option to pijuice_gui and after that will commit it, most likely to a new V1.8 branch

LG nach Wien

**Reply by lcnittl** [reply]
Awesome! Happy that I was able to provide some useful changes. Thanks again for the very quick implementation of this feature – looking forward to its official release.
Bis zum nächsten Mal!

**Reply by lcnittl** [reply]
Is there any news in this implementation - or I am overseeing the respective commits? If wanted, I can offer to file a PR (and also try to add the functionality to the gui tool). LG

**Comment by tvoverbeek** [comment]
Apologies for the delay. Had it ready to commit, but then other stuff happened and it got on the back burner.
Anyway, committed to the master branch now.
Thanks for the reminder.

**Comment by lcnittl** [comment]
Absolutely no problem :-D Thanks for the commit!
